ql/src/java/org/apache/hadoop/hive/ql/exec/vector/expressions/aggregates/VectorUDAFBloomFilterMerge.java

```diff
@@ -77,6 +75,211 @@ public void reset() {
+      int working = 0;
+      for (BloomFilterMergeWorker w : workers) {
+        if (w.isMerging.get()) {
+          working += 1;
+        }
+      }
+      return working;
+    }
+
+    private static void splitVectorAcrossWorkers(BloomFilterMergeWorker[] workers, byte[] bytes,
+        int start, int length) {
+      if (bytes == null || length == 0) {
+        return;
+      }
+      /*
+       * This will split a byte[] across workers as below:
+       * let's say there are 10 workers for 7813 bytes, in this case
+       * length: 7813, elementPerBatch: 781
+       * bytes assigned to workers: inclusive lower bound, exclusive upper bound
+       * 1. worker: 5 -> 786
+       * 2. worker: 786 -> 1567
+       * 3. worker: 1567 -> 2348
+       * 4. worker: 2348 -> 3129
+       * 5. worker: 3129 -> 3910
+       * 6. worker: 3910 -> 4691
+       * 7. worker: 4691 -> 5472
+       * 8. worker: 5472 -> 6253
+       * 9. worker: 6253 -> 7034
+       * 10. worker: 7034 -> 7813 (last element per batch is: 779)
+       *
+       * This way, a particular worker will be given with the same part
+       * of all bloom filters along with the shared base bloom filter,
+       * so the bitwise OR function will not be a subject of threading/sync issues.
+       */
+      int elementPerBatch =
+          (int) Math.ceil((double) (length - START_OF_SERIALIZED_LONGS) / workers.length);
+
+      for (int w = 0; w < workers.length; w++) {
+        int modifiedStart = START_OF_SERIALIZED_LONGS + w * elementPerBatch;
+        int modifiedLength = (w == workers.length - 1)
+          ? length - (START_OF_SERIALIZED_LONGS + w * elementPerBatch) : elementPerBatch;
+
+        ElementWrapper wrapper =
+            new ElementWrapper(bytes, start, length, modifiedStart, modifiedLength);
+        workers[w].add(wrapper);
+      }
+    }
+
+    public void shutdownAndWaitForMergeTasks() {
+      /**
+       * Executor.shutdownNow() is supposed to send Thread.interrupt to worker threads, and they are
+       * supposed to finish their work.
+       */
+      executor.shutdownNow();
+      try {
+        executor.awaitTermination(180, TimeUnit.SECONDS);
+      } catch (InterruptedException e) {
+        LOG.warn("Bloom filter merge is interrupted while waiting to finish, this is unexpected",
+            e);
+      }
+    }
+  }
+
+  private static class BloomFilterMergeWorker implements Runnable {
+    private BlockingQueue<ElementWrapper> queue;
+    private byte[] bfAggregation;
+    private int bfAggregationStart;
+    private int bfAggregationLength;
+    AtomicBoolean isMerging = new AtomicBoolean(false);
+
+    public BloomFilterMergeWorker(byte[] bfAggregation, int bfAggregationStart,
+        int bfAggregationLength) {
+      this.bfAggregation = bfAggregation;
+      this.bfAggregationStart = bfAggregationStart;
+      this.bfAggregationLength = bfAggregationLength;
+      this.queue = new ArrayBlockingQueue<>(VectorizedRowBatch.DEFAULT_SIZE * 2);
```

Comment:
Why VectorizedRowBatch.DEFAULT_SIZE * 2 aka 2048 ?
maybe workers.len * 2?

Reply:
not really, workers.len is horizontal, which is related to how many parts a bloom filter is split into, but queue.length needs to hold the same portion of the incoming bloom filters, so 1000 incoming bloom filters means 1000 items in the queue, so it's vertical (however, hopefully, ~1000 items won't be present in the queue if it's busy working, but I needed to find a safe value)...

Reply:
I see, is the BF num available to use here? Maybe BF num * 2 would be safe

Reply:
I'm afraid it's not available in advance, bloom filters comes one by one, in my testing scenario it is ~1200 row batches, and every row batch contains 1 bloom filter

Reply:
Its just not clear to me what this number depends on -- any chance this could greatly vary at a non-test scenario?
* If not VectorizedRowBatch.DEFAULT_SIZE * 2 is fine (we should just add a comment why this number was picked)
* If yes we could even have an unbounded queue here right?

Reply:
if there are 1000 upstream mapper tasks (creating bloom filters), there will be 1000 rowbatches (=1000 bloom filters), for example on TPCDS 30GB there were 1000<x<2000...anyway, you're absolutely right, I don't want to take care of correct bounds, which is unpredictable, I've just chosen a wrong implementation...I'm going to change this to LinkedBlockingDeque and letting this size confusion go